File: examples/dekita/.claude/hooks/pr_related_issue_check.py
#!/usr/bin/env python3
# - 責務: gh pr create 時に関連オープンIssueの確認を促す
# - 重複なし: closes-keyword-check.py はClosesキーワードの有無をチェック、本フックは関連Issue検索
# - 非ブロック型: 警告のみ、PRは作成可能
# - AGENTS.md: Issue #1849 に基づく実装
"""
Hook to warn about related open Issues when creating a PR.

When `gh pr create` is executed, this hook:
1. Extracts keywords from PR title and body
2. Searches for related open Issues using `gh issue list --search`
3. Shows a warning with related Issues (non-blocking)

Issue #1849: PR作成時に関連Issue確認を促すフック
"""

import json
import re
import subprocess
import sys

from lib.execution import log_hook_execution
from lib.session import parse_hook_input
from lib.strings import strip_quoted_strings

# Hook name for logging
HOOK_NAME = "pr-related-issue-check"

# Stop words to exclude from keyword extraction (English and Japanese)
STOP_WORDS = {
    # English
    "a",
    "an",
    "the",
    "is",
    "are",
    "was",
    "were",
    "be",
    "been",
    "being",
    "have",
    "has",
    "had",
    "do",
    "does",
    "did",
    "will",
    "would",
    "could",
    "should",
    "may",
    "might",
    "must",
    "shall",
    "can",
    "need",
    "and",
    "or",
    "but",
    "if",
    "then",
    "else",
    "when",
    "where",
    "why",
    "how",
    "what",
    "which",
    "who",
    "whom",
    "this",
    "that",
    "these",
    "those",
    "for",
    "with",
    "from",
    "into",
    "onto",
    "upon",
    "about",
    "after",
    "before",
    "above",
    "below",
    "between",
    "under",
    "over",
    "through",
    "during",
    "until",
    "while",
    "of",
    "at",
    "by",
    "in",
    "on",
    "to",
    "as",
    "it",
    "its",
    "not",
    "no",
    "yes",
    "all",
    "any",
    "both",
    "each",
    "few",
    "more",
    "most",
    "other",
    "some",
    "such",
    "only",
    "own",
    "same",
    "so",
    "than",
    "too",
    "very",
    "just",
    "also",
    "now",
    "new",
    # Japanese particles
    "を",
    "が",
    "に",
    "で",
    "は",
    "の",
    "と",
    "も",
    "や",
    "から",
    "まで",
    "より",
    "へ",
    "など",
    "か",
    "ね",
    "よ",
    "わ",
    # Common PR/Git words
    "fix",
    "feat",
    "feature",
    "add",
    "update",
    "remove",
    "delete",
    "change",
    "modify",
    "refactor",
    "improve",
    "bug",
    "issue",
    "pr",
    "pull",
    "request",
    "merge",
    "branch",
    "commit",
    "push",
    "test",
    "docs",
    "chore",
}

# Maximum number of keywords to extract
MAX_KEYWORDS = 5

# Maximum number of Issues to display
MAX_ISSUES_TO_DISPLAY = 5

# Minimum keyword length
MIN_KEYWORD_LENGTH = 3


def is_gh_pr_create_command(command: str) -> bool:
    """Check if command is a gh pr create command."""
    if not command.strip():
        return False
    stripped_command = strip_quoted_strings(command)
    return bool(re.search(r"gh\s+pr\s+create\b", stripped_command))


def extract_pr_title(command: str) -> str | None:
    """Extract the PR title from gh pr create command.

    Handles:
    - --title "..."
    - -t "..."
    - --title="..."

    Returns None if title is not explicitly specified.
    """
    # Standard patterns (ordered by specificity)
    dq_content = r'([^"\\]*(?:\\.[^"\\]*)*)'  # Double-quoted content with escapes
    sq_content = r"([^'\\]*(?:\\.[^'\\]*)*)"  # Single-quoted content with escapes
    patterns = [
        rf'--title="{dq_content}"',  # --title="..."
        rf"--title='{sq_content}'",  # --title='...'
        rf'-t="{dq_content}"',  # -t="..."
        rf"-t='{sq_content}'",  # -t='...'
        rf'--title\s+"{dq_content}"',  # --title "..."
        rf"--title\s+'{sq_content}'",  # --title '...'
        rf'-t\s+"{dq_content}"',  # -t "..."
        rf"-t\s+'{sq_content}'",  # -t '...'
    ]

    for pattern in patterns:
        match = re.search(pattern, command)
        if match:
            return match.group(1)

    return None


def extract_pr_body(command: str) -> str | None:
    """Extract the PR body from gh pr create command.

    Handles:
    - --body "..."
    - -b "..."
    - --body="..."
    - HEREDOC patterns: --body "$(cat <<'EOF' ... EOF)"

    Returns None if body is not explicitly specified inline.
    """
    # Try HEREDOC pattern first (most complex)
    heredoc_pattern = r'--body\s+"\$\(cat\s+<<[\'"]?(\w+)[\'"]?\s*(.*?)\s*\1\s*\)"'
    heredoc_match = re.search(heredoc_pattern, command, re.DOTALL)
    if heredoc_match:
        return heredoc_match.group(2)

    # Standard patterns (ordered by specificity)
    dq_content = r'([^"\\]*(?:\\.[^"\\]*)*)'  # Double-quoted content with escapes
    sq_content = r"([^'\\]*(?:\\.[^'\\]*)*)"  # Single-quoted content with escapes
    patterns = [
        rf'--body="{dq_content}"',  # --body="..."
        rf"--body='{sq_content}'",  # --body='...'
        rf'-b="{dq_content}"',  # -b="..."
        rf"-b='{sq_content}'",  # -b='...'
        rf'--body\s+"{dq_content}"',  # --body "..."
        rf"--body\s+'{sq_content}'",  # --body '...'
        rf'-b\s+"{dq_content}"',  # -b "..."
        rf"-b\s+'{sq_content}'",  # -b '...'
    ]

    for pattern in patterns:
        match = re.search(pattern, command)
        if match:
            return match.group(1)

    return None


def extract_keywords(title: str | None, body: str | None) -> list[str]:
    """Extract keywords from PR title and body.

    Args:
        title: PR title text
        body: PR body text

    Returns:
        List of keywords (max MAX_KEYWORDS), sorted by length descending
        to prioritize more specific terms.
    """
    text = ""
    if title:
        text += title + " "
    if body:
        text += body

    if not text.strip():
        return []

    # Extract words: alphanumeric sequences and Japanese characters
    words = re.findall(r"[a-zA-Z0-9\u3040-\u309f\u30a0-\u30ff\u4e00-\u9fff]+", text)

    # Filter words
    keywords = []
    seen = set()
    for word in words:
        word_lower = word.lower()
        # Skip if too short, is a stop word, or already seen
        if len(word) < MIN_KEYWORD_LENGTH:
            continue
        if word_lower in STOP_WORDS:
            continue
        if word_lower in seen:
            continue
        seen.add(word_lower)
        keywords.append(word)

    # Sort by length descending (longer words are more specific)
    keywords.sort(key=len, reverse=True)

    return keywords[:MAX_KEYWORDS]


def search_related_issues(keywords: list[str]) -> list[dict]:
    """Search for related open Issues using gh CLI.

    Args:
        keywords: List of keywords to search for

    Returns:
        List of Issue dicts with 'number' and 'title' keys.
    """
    if not keywords:
        return []

    # Build search query: OR-join keywords for broader match
    # GitHub search with spaces uses AND, so we explicitly use "OR" operator
    search_query = " OR ".join(keywords)

    try:
        result = subprocess.run(
            [
                "gh",
                "issue",
                "list",
                "--search",
                search_query,
                "--state",
                "open",
                "--limit",
                "10",
                "--json",
                "number,title",
            ],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if result.returncode != 0:
            return []

        issues = json.loads(result.stdout)
        return issues[:MAX_ISSUES_TO_DISPLAY]

    except (subprocess.TimeoutExpired, json.JSONDecodeError, OSError):
        return []


def main():
    """
    PreToolUse hook for Bash commands.

    Warns about related open Issues when creating a PR.
    """
    result = {"decision": "approve"}
    keywords_used = []

    try:
        data = parse_hook_input()
        tool_input = data.get("tool_input", {})
        command = tool_input.get("command", "")

        # Only check gh pr create commands
        if is_gh_pr_create_command(command):
            # Extract title and body
            title = extract_pr_title(command)
            body = extract_pr_body(command)

            # Extract keywords
            keywords = extract_keywords(title, body)
            keywords_used = keywords

            # Search for related Issues if keywords found
            if keywords:
                related_issues = search_related_issues(keywords)

                if related_issues:
                    # Format the warning message
                    issue_list = "\n".join(
                        f"  #{issue['number']}: {issue['title']}" for issue in related_issues
                    )
                    result["systemMessage"] = f"""⚠️ 関連するオープンIssueがあります

以下のIssueを確認しましたか？
{issue_list}

確認済みの場合は続行してください。

（検索キーワード: {", ".join(keywords)}）"""

                    if len(related_issues) >= MAX_ISSUES_TO_DISPLAY:
                        result["systemMessage"] += "\n\n💡 他にも関連Issueがある可能性があります。"

    except Exception as e:
        print(f"[{HOOK_NAME}] Hook error: {e}", file=sys.stderr)
        result = {"decision": "approve"}

    # Log execution
    log_hook_execution(
        HOOK_NAME,
        result.get("decision", "approve"),
        result.get("systemMessage"),
        {"keywords": keywords_used} if keywords_used else None,
    )
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/pr_related_issue_check.ts
#!/usr/bin/env bun
/**
 * gh pr create 時に関連オープンIssueの確認を促す。
 *
 * Why:
 *   PR作成時に関連するオープンIssueがあることを知らないと、
 *   重複作業や見落としが発生する。キーワードベースで関連Issueを
 *   検索し、確認を促す。
 *
 * What:
 *   - gh pr createコマンドからタイトル・ボディを抽出
 *   - キーワードを抽出してIssue検索
 *   - 関連するオープンIssueを警告表示
 *
 * Remarks:
 *   - 非ブロック型（警告のみ）
 *   - ストップワード（助詞、一般的なGit用語）を除外
 *   - 最大5件のIssueを表示
 *
 * Changelog:
 *   - silenvx/dekita#1849: フック追加
 *   - silenvx/dekita#3160: TypeScript移行
 */

import { extractPrBody, extractPrTitle } from "../lib/command";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";
import { asyncSpawn } from "../lib/spawn";
import { stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "pr-related-issue-check";

// =============================================================================
// Constants
// =============================================================================

// Stop words to exclude from keyword extraction
const STOP_WORDS = new Set([
  // English
  "a",
  "an",
  "the",
  "is",
  "are",
  "was",
  "were",
  "be",
  "been",
  "being",
  "have",
  "has",
  "had",
  "do",
  "does",
  "did",
  "will",
  "would",
  "could",
  "should",
  "may",
  "might",
  "must",
  "shall",
  "can",
  "need",
  "and",
  "or",
  "but",
  "if",
  "then",
  "else",
  "when",
  "where",
  "why",
  "how",
  "what",
  "which",
  "who",
  "whom",
  "this",
  "that",
  "these",
  "those",
  "for",
  "with",
  "from",
  "into",
  "onto",
  "upon",
  "about",
  "after",
  "before",
  "above",
  "below",
  "between",
  "under",
  "over",
  "through",
  "during",
  "until",
  "while",
  "of",
  "at",
  "by",
  "in",
  "on",
  "to",
  "as",
  "it",
  "its",
  "not",
  "no",
  "yes",
  "all",
  "any",
  "both",
  "each",
  "few",
  "more",
  "most",
  "other",
  "some",
  "such",
  "only",
  "own",
  "same",
  "so",
  "than",
  "too",
  "very",
  "just",
  "also",
  "now",
  "new",
  // Japanese particles
  "を",
  "が",
  "に",
  "で",
  "は",
  "の",
  "と",
  "も",
  "や",
  "から",
  "まで",
  "より",
  "へ",
  "など",
  "か",
  "ね",
  "よ",
  "わ",
  // Common PR/Git words
  "fix",
  "feat",
  "feature",
  "add",
  "update",
  "remove",
  "delete",
  "change",
  "modify",
  "refactor",
  "improve",
  "bug",
  "issue",
  "pr",
  "pull",
  "request",
  "merge",
  "branch",
  "commit",
  "push",
  "test",
  "docs",
  "chore",
]);

const MAX_KEYWORDS = 5;
const MAX_ISSUES_TO_DISPLAY = 5;
const MIN_KEYWORD_LENGTH = 3;

// =============================================================================
// Command Detection
// =============================================================================

/**
 * Check if command is a gh pr create command.
 */
function isGhPrCreateCommand(command: string): boolean {
  if (!command.trim()) {
    return false;
  }
  const stripped = stripQuotedStrings(command);
  return /gh\s+pr\s+create\b/.test(stripped);
}

// extractPrTitle and extractPrBody are imported from ../lib/command

// =============================================================================
// Keyword Extraction
// =============================================================================

/**
 * Extract keywords from PR title and body.
 */
function extractKeywords(title: string | null, body: string | null): string[] {
  let text = "";
  if (title) {
    text += `${title} `;
  }
  if (body) {
    text += body;
  }

  if (!text.trim()) {
    return [];
  }

  // Extract words: alphanumeric and Japanese characters
  const words = text.match(/[a-zA-Z0-9\u3040-\u309f\u30a0-\u30ff\u4e00-\u9fff]+/g) ?? [];

  // Filter words
  const keywords: string[] = [];
  const seen = new Set<string>();

  for (const word of words) {
    const wordLower = word.toLowerCase();

    // Skip if too short, is a stop word, or already seen
    if (word.length < MIN_KEYWORD_LENGTH) {
      continue;
    }
    if (STOP_WORDS.has(wordLower)) {
      continue;
    }
    if (seen.has(wordLower)) {
      continue;
    }

    seen.add(wordLower);
    keywords.push(word);
  }

  // Sort by length descending (longer words are more specific)
  keywords.sort((a, b) => b.length - a.length);

  return keywords.slice(0, MAX_KEYWORDS);
}

// =============================================================================
// Issue Search
// =============================================================================

interface RelatedIssue {
  number: number;
  title: string;
}

/**
 * Search for related open Issues using gh CLI.
 */
async function searchRelatedIssues(keywords: string[]): Promise<RelatedIssue[]> {
  if (keywords.length === 0) {
    return [];
  }

  // Build search query with OR-join
  const searchQuery = keywords.join(" OR ");

  try {
    const result = await asyncSpawn(
      "gh",
      [
        "issue",
        "list",
        "--search",
        searchQuery,
        "--state",
        "open",
        "--limit",
        "10",
        "--json",
        "number,title",
      ],
      { timeout: TIMEOUT_MEDIUM * 1000 },
    );

    if (!result.success) {
      return [];
    }

    const issues = JSON.parse(result.stdout);
    return issues.slice(0, MAX_ISSUES_TO_DISPLAY);
  } catch {
    return [];
  }
}

// =============================================================================
// Main
// =============================================================================

async function main(): Promise<void> {
  const result: { decision?: string; systemMessage?: string } = {};
  let keywordsUsed: string[] = [];
  let sessionId: string | undefined;

  try {
    const input = await parseHookInput();
    sessionId = input.session_id;
    const toolInput = input.tool_input as Record<string, unknown> | undefined;
    const command = (toolInput?.command as string) ?? "";

    if (isGhPrCreateCommand(command)) {
      // Extract title and body
      const title = extractPrTitle(command);
      const body = extractPrBody(command);

      // Extract keywords
      const keywords = extractKeywords(title, body);
      keywordsUsed = keywords;

      // Search for related Issues if keywords found
      if (keywords.length > 0) {
        const relatedIssues = await searchRelatedIssues(keywords);

        if (relatedIssues.length > 0) {
          const issueList = relatedIssues
            .map((issue) => `  #${issue.number}: ${issue.title}`)
            .join("\n");

          result.systemMessage = `⚠️ 関連するオープンIssueがあります

以下のIssueを確認しましたか？
${issueList}

確認済みの場合は続行してください。

（検索キーワード: ${keywords.join(", ")}）`;

          if (relatedIssues.length >= MAX_ISSUES_TO_DISPLAY) {
            result.systemMessage += "\n\n💡 他にも関連Issueがある可能性があります。";
          }
        }
      }
    }
  } catch (error) {
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(error)}`);
  }

  // Log execution
  await logHookExecution(
    HOOK_NAME,
    result.decision ?? "approve",
    result.systemMessage,
    keywordsUsed.length > 0 ? { keywords: keywordsUsed } : undefined,
    { sessionId },
  );

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
